aten/src/ATen/native/cuda/UniqueCub.cu

```diff
@@ -0,0 +1,201 @@
+#include <ATen/native/cuda/UniqueCub.cuh>
+
+#include <ATen/cuda/CUDAContext.h>
+#include <ATen/cuda/CUDAApplyUtils.cuh>
+#include <ATen/cuda/cub.cuh>
+
+namespace at {
+namespace native {
+namespace internal {
+
+namespace {
+
+template <typename scalar_t>
+__global__ void adjacent_difference_kernel(
+    int64_t n,
+    const scalar_t* input,
+    int* output) {
+  auto begin = blockIdx.x * blockDim.x + threadIdx.x;
+  auto stride = gridDim.x * blockDim.x;
+  for (auto i = begin; i < n; i += stride) {
+    output[i] = i > 0 ? input[i] != input[i - 1] : 0;
+  }
+}
+
+__global__ void scatter_kernel(
+    int64_t n,
+    const int64_t* input,
+    const int64_t* indices,
+    int64_t* output) {
+  auto begin = blockIdx.x * blockDim.x + threadIdx.x;
+  auto stride = gridDim.x * blockDim.x;
+  for (auto i = begin; i < n; i += stride) {
+    output[indices[i]] = input[i];
+  }
+}
+
+// A variation of compute_unique (defined in Unique.cu) that doesn't allow
+// customizing equal and not_equal (needed to use CUB).
+template <typename scalar_t>
+std::tuple<Tensor, Tensor, Tensor, int64_t> compute_unique(
+    const Tensor& sorted,
+    const Tensor& sorted_indices,
+    const bool return_inverse,
+    const bool return_counts) {
+  int64_t num_inp = sorted.numel();
+  TORCH_CHECK(
+      num_inp <= INT_MAX, "num_inp ", num_inp, " is too big to for CUB");
+  auto options = sorted.options().dtype(kLong);
+  const scalar_t* data = sorted.data_ptr<scalar_t>();
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+
+  // inverse indices
+  Tensor inverse_indices;
+  if (!return_inverse || num_inp == 0) {
+    inverse_indices = at::empty({0}, options);
+  } else {
+    TORCH_CHECK(
+        sorted_indices.defined(),
+        "return_inverse is set to true, but sorted_indices is undefined. Send a bug report!");
+    inverse_indices = at::empty({num_inp}, options);
+    int* inv_loc_ptr =
+        reinterpret_cast<int*>(inverse_indices.data_ptr<int64_t>());
```

Comment:
Why casting the pointer to `int*`? Is this correct?

Reply:
If you want to use int, why not creating a `kInt` tensor since the beginning?

Reply:
This is a micro-optimization to reuse inverse_indices instead of creating a new one. adjacent_difference_kernel just needs int (actually boolean suffices) but scatter_kernel outputs int64. If you think this is a premature optimization, I'm happy to create a new tensor here :)

Reply:
I see, this is a good optimization. You can keep it.